release-19.1: distsqlpb: whitelist node unavailability errors

knz opened the pull request:
Backport 2/2 commits from #37367.
Backport 1/1 commits from #37800.

/cc @cockroachdb/release

---

Fixes #37215.

A node being down during distsql query processing is a legitimate (and
expected) error. It needs not be reported to telemetry.

knz committed:
distsqlpb: avoid crashing a node upon unknown error details

Prior to this patch, a distsql gateway would crash if it received an
error payload of a type it didn't know about. This is unfair to the
user, as an error (regardless of payload) is just an error.

This patch removes the panic and produces a valid error (with a Sentry
report, so we can investigate further).

Release note: None

knz committed:
distsqlpb: whitelist node unavailability errors

A node being down during distsql query processing is a legitimate (and
expected) error. It needs not be reported to telemetry.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37789)
<!-- Reviewable:end -->

knz committed:
sql: fix schema change auto-retry upon node failures

Prior to #37367, a node unavailable error was reported in distsql as
a pgerror with code "internal" (assertion). Change #37367 changes this
to report node availability using a different code.

Meanwhile, the schema change logic wants to be able to retry if a
schema change appears to fail due a node going down. Since this is not
exercised in CI (only in nightly test), #37367 forgot about that. This
commit completes the fix.

(Note that this dance with error codes are band-aids; a more robust
fix is upcoming in #37765 and following.)

Release note: None

andreimatei reviewed (approved):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37789#-:-Lfz7G0d3V7ooJMmdToT:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->
